File: C/5/6.c
#include <stdio.h>
#include <string.h>

#define MAXSTR 1000

/**
 * @brief: Read a line into s
 *
 * @param s     Pointer to character array where line will be saved
 * @param lim   Max number of characters read per line
 *
 * @return int  Length of s
 */
int getline(char *s, int lim)
{
    char *aux = s;
    int c, n = 0;

    while (n < lim && (c = getchar()) != EOF && c != '\n') {
        *s++ = c;
        n += 1;
    }

    *s = '\0';

    return s - aux;
}

/**
 * @brief Reverses string s in place.
 *
 * @param s String to be reversed.
 */
void reverse(char *s)
{
    int c = 0, size = strlen(s);
    char *end = s + size - 1;

    while (s < end)
    {
        c = *s;
        *s++ = *end;
        *end-- = c;
    }
}

/**
 * @brief A function to check for the rightmost occurence of t in s.
 *
 * @param char* s   The string to be searched at
 * @param char* t   The pattern to be searched for
 *
 * @return int      Position of rightmost occurrence .
 */
int strindex(char *s, char *t)
{
    char *s_aux = s;
    char *t_aux = t;

    int i = 0, rightmost = -1;

    for (i = 0; *s != '\0'; s++, i++) {
        for (s_aux = s, t_aux = t; *t_aux != '\0' && *s_aux==*t_aux; s_aux++, t_aux++);

        if (t_aux > t && *t_aux == '\0') {
            rightmost = i;
        }

    }

    return rightmost;
}

main ()
{
    int n = 0;
    char input[MAXSTR];

    char numbers[] = "12345";

    printf("Enter a line to be read:\n");

    n = getline(input, MAXSTR);
    printf("%d characters were read:\t\"%s\"\n", n, input);

    reverse(input);
    printf("Check it reversed:\t\"%s\"\n", input);

    reverse(input);
    printf("Check it reversed back:\t\"%s\"\n", input);

    printf("The rightmost index of \"%s\" inside \"%s\" is %d\n", numbers, input, strindex(input, numbers));

}

File: rustlang/5/6.rs
use std::io;

/// Reverses string s in place.
///
/// @param s String to be reversed.
fn reverse(s: &mut String)
{
    let size = s.chars().count();

    for _i in 0..size {
        let _c = s.pop().unwrap_or('#');
        s.insert(_i as usize, _c);
    }
}

/// A function to check for the rightmost occurence of t in s.
///
/// @param s         The string to be searched at
/// @param t         The pattern to be searched for
///
/// @return usize    Position of rightmost occurrence .
fn strindex(s: &str, t: &str) -> i32
{
    let mut rightmost: i32 = -1;


    for _idx in 0..=s.chars().count() {

        let mut s_iter = s.chars().skip(_idx);
        let mut t_iter = t.chars();

        loop {
            let _s = s_iter.next();
            let _t = t_iter.next();

            if _s == None || _s != _t { break; }
        }

        if t_iter.next() == None  {
            rightmost = _idx as i32;
        }
    }

    rightmost
}

fn main ()
{
    let mut input = String::new();

    let numbers = String::from("12345");

    print!("Enter a line to be read:\n");

    match io::stdin().read_line(&mut input) {
        Ok (n) => {
            input = input.trim_end().to_string();

            print!("{} characters were read:\t\"{}\"\n", n-1, input);

            reverse(&mut input);
            print!("Check it reversed:\t\"{}\"\n", input);

            reverse(&mut input);
            print!("Check it reversed back:\t\"{}\"\n", input);

            print!("The rightmost index of \"{}\" inside \"{}\" is {}\n", numbers, input, strindex(&input, &numbers));
        },
        Err(_) => {
            panic!("Unexpected error.");
        }
    }


}
